File: common/jwt-core/src/main/java/com/halcyon/tinder/jwtcore/JwtAuthenticationFilter.java
package com.halcyon.tinder.jwtcore;

import io.jsonwebtoken.Claims;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

@Component
@RequiredArgsConstructor
public class JwtAuthenticationFilter extends OncePerRequestFilter {

    private final JwtProvider jwtProvider;
    private final TokenRevocationService tokenRevocationService;

    private static final String AUTHORIZATION_HEADER = "Authorization";
    private static final String BEARER_PREFIX = "Bearer ";

    @Override
    protected void doFilterInternal(@NonNull HttpServletRequest request,
                                    @NonNull HttpServletResponse response,
                                    @NonNull FilterChain filterChain)
            throws ServletException, IOException {
        String jwtToken = getTokenFromRequest(request);
        boolean isRefresh = false;

        if (jwtToken != null && jwtProvider.isValidToken(jwtToken, isRefresh) && !tokenRevocationService.isRevoked(jwtToken)) {
            Claims claims = jwtProvider.extractAllClaims(jwtToken, isRefresh);

            var jwtAuthentication = new JwtAuthentication(true, claims.getSubject(), jwtToken);
            SecurityContextHolder.getContext().setAuthentication(jwtAuthentication);
        }

        filterChain.doFilter(request, response);
    }

    private String getTokenFromRequest(HttpServletRequest request) {
        String authHeader = request.getHeader(AUTHORIZATION_HEADER);

        if (authHeader != null && authHeader.startsWith(BEARER_PREFIX)) {
            return authHeader.substring(7);
        }

        return null;
    }
}

File: common/jwt-core/src/main/kotlin/com/halcyon/tinder/jwtcore/JwtAuthenticationFilter.kt
package com.halcyon.tinder.jwtcore

import io.jsonwebtoken.Claims
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
class JwtAuthenticationFilter(
    private val jwtProvider: JwtProvider,
    private val tokenRevocationService: TokenRevocationService
) : OncePerRequestFilter() {

    companion object {
        private const val AUTHORIZATION_HEADER = "Authorization"
        private const val BEARER_PREFIX = "Bearer "
    }

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        val jwtToken = getTokenFromRequest(request)
        val isRefresh = false

        if (jwtToken != null && jwtProvider.isValidToken(jwtToken, isRefresh) && !tokenRevocationService.isRevoked(jwtToken)) {
            val claims: Claims = jwtProvider.extractAllClaims(jwtToken, isRefresh)
            val jwtAuthentication = JwtAuthentication(true, claims.subject, jwtToken)

            SecurityContextHolder.getContext().authentication = jwtAuthentication
        }

        filterChain.doFilter(request, response)
    }

    private fun getTokenFromRequest(request: HttpServletRequest): String? {
        val authHeader = request.getHeader(AUTHORIZATION_HEADER)

        return if (authHeader != null && authHeader.startsWith(BEARER_PREFIX)) {
            authHeader.substring(BEARER_PREFIX.length)
        } else {
            null
        }
    }
}